oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -184,18 +184,44 @@ Imposing Constraints on Dependencies
 Although we usually want to use the latest available version of our
 dependencies in order to take advantage of the latest bug fixes, performance
 improvements, and security fixes, we sometimes need to impose some constraints
-on the version to be used.  These should be collected in
-``requirements/constraints.txt`` so they can be imposed uniformly across all
-the repository's requirements files; this is done via a ``-c constraints.txt``
-line just under the summary comment of each ``*.in`` file in the
-``requirements`` directory.  Some guidelines to keep in mind when populating
-this file:
+on the version to be used. These constraints can either be placed directly on
+the existing dependency in the ``*.in`` file or as a new entry in
+``requirements/constraints.txt``, according to the following rules:
+
+* If it is a constraint on a dependency listed in **a library's base.in**,
+  the constraint should be added to the existing dependency line.
+
+  * This allows constraints to be advertised via ``setup.py`` and taken
+    into account in dependency resolution of relying packages.
+  * If a package advertises an extra feature, constraints should be similarly
+    placed in the extra's corresponding ``*.in`` file.
+
+* All other constraints should instead be added to
+  ``requirements/constraints.txt``.
+
+  * The ``*.in`` files must include the line ``-c constraints.txt`` just below
+    their summary comment. This allows constraints to be shared uniformly
+    across multiple files. (The ``-c`` flag keeps the constraints from being
+    interpreted as dependencies.)
+  * Most constraints could be listed either here or in ``*.in`` files directly,
+    but the constraints file was historically preferred and so continuing to
+    prefer it reduces churn. The only case where the constraints file is
+    specifically required is when multiple ``*.in`` files declare the same
+    dependency, but in such a situation it is generally preferred to rearrange
+    things so that one ``*.in`` pulls in the other's ``*.txt`` file  using a
+    ``-r`` line, and only list the dependency in the latter.
+
+Some guidelines to keep in mind when adding constraints:
 
 * Version constraints should only be used to exclude dependency versions which
   are known (or strongly suspected) to not work in at least one context.
 * Constraints on indirect dependencies (used by dependencies but not directly
   by the code in the repository itself) can be added if needed to enforce a
-  compatible version.
+  compatible version. If these indirect dependencies are relevant to the base
+  dependencies in a library, then the dependency should be made explicit in
+  ``base.in`` along with a constraint and a comment explaining that it is only
+  listed for the constraint, and should be removed entirely once the constraint
+  is no longer needed.
```

Comment:
Should we use the term "maximum or pinned constraint" here, to differentiate from min constraints?

Reply:
Yes, I think better to express it here than as an exception later.